### PR TITLE
ScriptActorInterfaces, unbind on actor destroy.

### PR DESCRIPTION
**Pull request by anvilvapre**
Do not recreate all actor commands again - that should remain after actor destroy - using reflection. Rather remove the commands that should not remain. The trade off between reflection and object allocations - of both objects and members - versus removing many values from a dictionary. It seems that from the 20+ commands/members only 5-10 are left after an actor was destroyed.

- Add `Unbind` that removes all members for a command type.
- Add helper `CreateObjects`. 
- Remove some LINQ. Change argument to array. 

Note: it seems that a `ScriptActorInterfaces` is created lazily from `Actor`. At times - i.e. when switching maps - the `luaInterface` is created after the actor is already destroyed - only to after call `OnActorDestroyed`. Is there still a reason to do create the luaInterface? Or could we check `luaInterface.IsValueCreated`? Are rhere other cases where an actor starts destroyed?
https://github.com/OpenRA/OpenRA/blob/8dec998d8ff8d6cf5f357cdfbbd5630900d85f78/OpenRA.Game/Actor.cs#L423

Question: `ScriptMemberWrapper.Get` when `IsMethod == true` calls `CreateFunctionFromDelegate`. Couldn't this be done only of the first call after which in successive calls a cached value is returned? i.e. by copying the LuaFunction value using `CopyReference`? I assume the same functions are called often in scripts.
https://github.com/OpenRA/OpenRA/blob/8dec998d8ff8d6cf5f357cdfbbd5630900d85f78/OpenRA.Game/Scripting/ScriptMemberWrapper.cs#L100-L103

Question: in the `TryGetClrValue` function first an attempt is made to get a clr object using `TryGetClrObject(..)`. Even when this returns true - all other types are still tried after. Is that because the type of the out variable should always match? 
https://github.com/OpenRA/OpenRA/blob/8dec998d8ff8d6cf5f357cdfbbd5630900d85f78/OpenRA.Game/Scripting/ScriptTypes.cs#L43-L50

### CODE REVIEW

File: OpenRA.Game/Scripting/ScriptActorInterface.cs [hunk 1]
```diff
@@ -38,26 +37,21 @@ public ScriptActorInterface(ScriptContext context, Actor actor)
 
 		void InitializeBindings()
 		{
-			var commandClasses = Context.ActorCommands[actor.Info].AsEnumerable();
+			var commandClasses = Context.ActorCommands[actor.Info];
 
-			// Destroyed actors cannot have their traits queried
+			// Destroyed actors cannot have their traits queried. In rare casses actor may have already been destroyed.
```
Comment (abcdefg30) [inline OpenRA.Game/Scripting/ScriptActorInterface.cs]:
```suggestion
			// Destroyed actors cannot have their traits queried. In rare cases the actor may have already been destroyed.
```

File: OpenRA.Game/Scripting/ScriptObjectWrapper.cs [hunk 2]
```diff
@@ -44,6 +61,13 @@ protected void Bind(IEnumerable<object> clrObjects)
 			}
 		}
 
+		protected void Unbind(Type targetType)
+		{
+			foreach (var m in members)
+				if (targetType == m.Value.Target.GetType())
+					members.Remove(m.Key);
```
Comment (RoosterDragon) [inline OpenRA.Game/Scripting/ScriptObjectWrapper.cs:68]: I was going to note that `Remove` inside a foreach would be invalid, but apparently that's ok for dictionary these days. Neat.

**Review by RoosterDragon — DISMISSED** [review]

**Comment by abcdefg30** [comment]
The commits should be squashed.

**Review by RoosterDragon — APPROVED** [review]

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/69e5d07956a3046f44e5a51b6c9311a5a647b2a6...4b17d1f66a9e52776903e016e9a2010b70728491)

File: OpenRA.Game/Scripting/ScriptObjectWrapper.cs (hunk 2 above)
Comment (nevercast) [reply OpenRA.Game/Scripting/ScriptObjectWrapper.cs:68]: #20355 not on mono it seems

**Pull request merged**